Unsaved changes notification on review form

bickelj opened the pull request:
Fixes #3977 

## Test Steps

- As Reviewer, add or edit a review on a submission (example `/apply/submissions/2/reviews/3/`)
- Change one of the fields without clicking Submit
- Close the tab: a notification should require confirmation
- Click back: a notification should require confirmation
- Click another tab: a notification should require confirmation

gmurtaza00 committed:
Unsaved changes notification on review form

Issue #3977

theskumar commented on hypha/static_src/javascript/review-form-actions.js:
```diff
@@ -0,0 +1,44 @@
+            });
+            $(".form--score-box").text("Score: " + score);
+        }
+    });
```
This piece of code doesn't seems to be related to form alert, if so, I would better to have it in another PR.

bickelj replied:
Whoops! Correct! That is in https://github.com/HyphaApp/hypha/pull/3964. I removed it here.

theskumar commented on hypha/static_src/javascript/review-form-actions.js:
```diff
@@ -0,0 +1,44 @@
+(function ($) {
+    "use strict";
+    $(document).ready(function () {
+        var form = $("form");
```
I would suggest using a specific selector instead of putting the selector all the forms. 

Modify the markup to look something like: 

```html
<section class="review-form">
    <!-- rendered form. -->
</section>
```

and then use the selector `".review-form form"` 

If you able to modify the `<form>` tag itself then something like `"form [data-alert-if-unsaved]"` for an html that might look like: 

```html
<form data-alert-if-unsaved action="..." method="...">
</form>
```

bickelj replied:
I took this one step further and used an `id` on the form. That way it is exactly the form we care about.

frjo commented on hypha/static_src/javascript/review-form-actions.js:
```diff
@@ -0,0 +1,44 @@
+            $(".form--score-box").text("Score: " + score);
+        }
+    });
+})(jQuery);
```
Please also rewrite to vanillajs or alpinejs. We are moving away from jQuery.

bickelj replied:
Done.

bickelj committed:
Refactor reviewer unsaved changes notification

Based on feedback, refactor the reviewer unsaved changes notification
to avoid deprecated jQuery, use a precise selector for the form
element, and remove unrelated code.

Credit to dolphin-mixtral:8x7b cfada4ba31c7 on ollama 0.3.7-rc5, to
Saurabh Kumar for the selector suggestion, and to Fredrik Jonsson for
finding a miss.

Issue #3977

frjo commented on hypha/apply/review/templates/review/review_edit_form.html:
```diff
@@ -13,7 +13,7 @@
 
     <div class="wrapper wrapper--medium wrapper--inner-space-medium">
-        <form class="form form--with-p-tags form--scoreable" action="" method="post" novalidate>
+        <form id="review-form-edit" class="form form--with-p-tags form--scoreable" action="" method="post" novalidate>
```
This id needs to be added to `hypha/apply/review/templates/review/review_form.html` as well.

bickelj replied:
Whoops, nice catch, added!

frjo reviewed (changes requested):
A minor fix and this is ready to go to testing and then being merged I think.

bickelj commented:
b9f89ca rebased on main and included one change to `hypha/apply/review/templates/review/review_form.html` as suggested by @frjo.

bickelj commented:
@frjo Can this be merged in the next iteration?

frjo commented:
Adding it to the test site now. That makes it a step closer.

frjo commented:
@bickelj Please rebase to fix merge conflict. Would also be easier if moved to Hypha repo, thanks.

bickelj commented:
Superseded by #4176 where it has been rebased and pushed to a branch accessible to the Hypha team.
